### PR TITLE
Add POCO read materialization

### PR DESCRIPTION
**Pull request by alex-clickhouse**
## Summary

Adds a streaming POCO read path for `IClickHouseClient` and `ClickHouseDataReader`, backed by a per-client `PocoTypeRegistry` shared with the existing binary-insert path. Strict v1 type-matching: no implicit conversions, with rich diagnostics on mismatch.

## API

- `RegisterBinaryInsertType<T>()` — existing, unchanged. Insert-only. Kept for backwards compatibility (to be obsoleted in a later release).
- `RegisterPocoType<T>()` — new. Canonical registration: sets up both insert and read mappings. Both mappings are validated up front; if either validation throws, neither mapping is committed, so the registry is left untouched.
- `client.QueryAsync<T>(sql, parameters?, options?, ct?)` — `IAsyncEnumerable<T>`. Streams rows lazily; the underlying reader is disposed on completion, fault, or early break.
- `reader.MapTo<T>()` — materializes the current row into a fresh `T` without advancing the reader. Pairs with `Read()` for ADO.NET-style consumers.

`[ClickHouseColumn(Name = ...)]` aliases and `[ClickHouseNotMapped]` exclusion behave the same as on the existing write path.

## How it works

1. **Registration** compiles getters (insert) and setters + a raw-`newobj` constructor (read) via `Expression.Lambda` and caches them in a per-client `ConcurrentDictionary`. The compiled constructor uses raw `newobj`, so types with `required` properties are supported.
2. **Headers parse** when the reader opens. The first `MapTo<T>` (or first iteration of `QueryAsync<T>`) builds an ordinal-ordered binding plan from `(column ordinal, property index)` pairs and **fail-fasts** on impossible mappings: each column's `FrameworkType` is checked against the target property type before any rows are materialized. Polymorphic columns (`FrameworkType == typeof(object)` — Variant/Dynamic/JSON/Object) skip the static check.
3. **Per-row**, the binding plan drives the loop: null/`DBNull` is gated by `CanAssignNull`; otherwise the compiled boxed setter is invoked. An `InvalidCastException` from the setter (only reachable for polymorphic columns or stale mismatches) is caught and rethrown as `InvalidOperationException` with column name, ClickHouse type, target property, and returned CLR type.

`ClickHouseDataReader.Read()` is unchanged. Existing readers keep the buffered `CurrentRow` available for arbitrary ordinal access alongside `MapTo<T>`.

## Type-matching rules (strict v1)

- **Case-sensitive** (`StringComparer.Ordinal`) column-to-property matching.
- **Missing columns** in the result leave properties at their default value.
- **Extra columns** in the result are ignored.
- **No implicit conversions**: no widening (`Int8` → `Int32`), no enum coercion (`Enum8` returns `string`, not your CLR enum), no `ClickHouseDecimal` ↔ `decimal` interop. Mismatches throw `InvalidOperationException` naming the POCO type, property, column, and returned CLR type.
- **Static mismatches fail fast** at the first `MapTo<T>` / iteration — before any rows are read.

## Limitations

- **Read registration requires a public parameterless constructor.** Records with primary constructors are rejected on the read side (still fine for `RegisterBinaryInsertType`, which doesn't construct).
- **Init-only, read-only, and non-public-setter properties** are silently skipped on the read side and keep their default value even if a matching column is present. The type still registers if at least one property has a public non-init setter.
- **`required` properties** are supported (the compiled constructor bypasses required-member enforcement via raw `newobj`).
- **Per-reader `bindingPlanCache` is a non-concurrent `Dictionary`** — fine because `DbDataReader` isn't thread-safe by spec; concurrent `MapTo<T>` calls on a single reader are not supported.
- **No conversion hooks in v1.** Future opt-in widening / enum / decimal coercions would be additive (a `PocoReadOptions` flag or `[ClickHouseColumn(Convert = true)]` attribute), with no public-API break.

## Performance

`PocoReadBenchmark` reads `system.numbers` projecting `(Id, Name, Value)` and compares manual `GetValue` row construction against `QueryAsync<T>` POCO materialization.

| Method         | Count  | Mean     | Ratio | Allocated |
|----------------|--------|---------:|------:|----------:|
| ManualGetValue | 100k   | 104.0 ms | 1.00  | 12.73 MB  |
| Poco           | 100k   | 115.1 ms | 1.14  | 12.73 MB  |
| ManualGetValue | 500k   | 327.0 ms | 1.00  | 61.55 MB  |
| Poco           | 500k   | 341.0 ms | 1.05  | 61.55 MB  |

Allocations are identical — column values are boxed once during reader buffering and reused. The remaining per-row overhead is the boxed `Action<T, object>` setter invocation; a future generic-typed-setter pass could close it.

Run on .NET 10 / Linux Ubuntu 22.04 / Intel Core Ultra 7 255U.

## Tests

Coverage spans registration edge cases (records, init-only, abstract, write-only, indexers, statics, fields, required, inheritance), the strict-v1 read matrix (alias, missing/extra, case sensitivity, nullable/non-nullable, decimal with and without `UseCustomDecimals`, type mismatch, unregistered, before-`Read`, after-end-of-stream, derived, multiple types per reader, parameters, query options), runtime skip of init-only/read-only/private-setter properties, plus review-driven additions:
- atomicity of `RegisterPocoType<T>` when either side's validation throws,
- `QueryAsync<T>` cancellation mid-iteration with a successful follow-up query,
- mid-stream server error surfacing as `ClickHouseServerException` through `QueryAsync<T>` (gated on ClickHouse 25.11+),
- full nullable-property matrix (`Nullable(Int32)` null, `Nullable(Int32)` non-null, plain `Int32` → `int?`).

A new example, `examples/Select/Select_005_PocoSelect.cs`, demonstrates the calling shapes end-to-end.

---

> [!NOTE]
> **Medium Risk**
> New public API on the hot query/read path with strict mapping rules that can break callers who expect widening or silent coercion; registry atomicity and reader lifecycle changes are well-tested but affect all POCO insert users via the shared registry refactor.
> 
> **Overview**
> This PR adds **POCO read materialization** alongside the existing binary-insert POCO path. Query results can be mapped into registered classes with **`QueryAsync<T>`** (lazy `IAsyncEnumerable`, reader disposed when enumeration ends) or **`ClickHouseDataReader.MapTo<T>()`** on the current row without advancing the reader.
> 
> **Registration** is centralized in a per-client **`PocoTypeRegistry`** (replacing the insert-only registry): **`RegisterPocoType<T>()`** validates and registers **both** insert and read mappings atomically; **`RegisterBinaryInsertType<T>()`** remains insert-only. **`ClickHouseConnection.RegisterPocoType<T>()`** forwards to the client so ADO.NET readers get the same registry. Readers created from **`ExecuteReaderAsync`** / commands now receive the registry for `MapTo`.
> 
> Read mapping uses **case-sensitive** column names, **`[ClickHouseColumn]`** / **`[ClickHouseNotMapped]`** parity with inserts, **no implicit conversions** (static assignability checked when the binding plan is built; mismatches throw **`InvalidOperationException`** with column/property context). **`Read()`** tracks **`hasCurrentRow`** so `MapTo` fails clearly before/after the stream.
> 
> Docs/changelog describe POCO reads under v1.3.0; tests, a benchmark, and **`Select_006_PocoSelect`** example cover registration edge cases, strict typing, and mid-stream server errors through **`QueryAsync`**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 80e0230a0251e5299a17d15c48f15de3425d1281. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: ClickHouse.Driver/IClickHouseClient.cs [hunk 1]
```diff
@@ -165,11 +165,43 @@ Task<long> InsertBinaryAsync<T>(
...
+    /// Registers a POCO type for both binary insert and read materialization. Types must be
+    /// registered before they can be inserted via <see cref="InsertBinaryAsync{T}"/> or
+    /// materialized via <see cref="QueryAsync{T}"/> / <see cref="ClickHouseDataReader.MapTo{T}"/>.
+    /// Both mappings are validated up front; if either validation throws, neither mapping is
+    /// committed and the registry is left untouched.
+    /// On the read side, init-only, read-only, and non-public-setter properties are silently
+    /// skipped — they keep their default value even if a matching result column is present.
+    /// </summary>
+    /// <typeparam name="T">The POCO type to register.</typeparam>
+    void RegisterPocoType<T>()
```
Comment (DanielBunting) [inline ClickHouse.Driver/IClickHouseClient.cs:186]: Small thing - since this adds members to `IClickHouseClient`, downstream implementors will need to add the new members on upgrade, which is source-breaking for them (eg. custom wrappers for logging/retry/routing). 

Could you potentially add something along the lines of: 
```
void RegisterPocoType<T>() where T : class =>
    throw new NotSupportedException(
        $"{GetType().Name} does not implement RegisterPocoType<T>. " +
        "Added in v1.4.0 — update your implementation or use ClickHouseClient.");
```

Existing implementations keep compiling and wrappers get a clear runtime error until they upgrade.

File: ClickHouse.Driver/IClickHouseClient.cs [hunk 2]
```diff
@@ -165,11 +165,43 @@ Task<long> InsertBinaryAsync<T>(
...
     void RegisterBinaryInsertType<T>()
         where T : class;
 
+    /// <summary>
+    /// Registers a POCO type for both binary insert and read materialization. Types must be
+    /// registered before they can be inserted via <see cref="InsertBinaryAsync{T}"/> or
+    /// materialized via <see cref="QueryAsync{T}"/> / <see cref="ClickHouseDataReader.MapTo{T}"/>.
+    /// Both mappings are validated up front; if either validation throws, neither mapping is
+    /// committed and the registry is left untouched.
+    /// On the read side, init-only, read-only, and non-public-setter properties are silently
```
Comment (DanielBunting) [inline ClickHouse.Driver/IClickHouseClient.cs:182]: Worth considering whether this should be opt-in rather than the default. The 'silent skip' failure mode is hard to diagnose; registration succeeds, queries run, and every property comes back at its CLR default with no exception, no log, nothing to help diagnose. 

Could RegisterPocoType<T> fail registration if the POCO has read-unmappable properties (init-only, read-only, non-public-setter) by default, with a `PocoReadOptions` type with a bool opt-out for users who genuinely want the current skip behaviour?

**Review by mzitnik — APPROVED** [review]

File: ClickHouse.Driver/IClickHouseClient.cs (hunk 2 above)
Comment (alex-clickhouse) [reply ClickHouse.Driver/IClickHouseClient.cs:182]: Didn't want to add another option for this, but there is now logging in place to make debugging poco registrations easy.
